### PR TITLE
Shooter code

### PR DESCRIPTION
**Pull request by elizaBarnett**

### CODE REVIEW

**Review by nbgray — APPROVED** [review]
Reviewed in person

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    private double maxLaunchSpeed = maxTangentialSpeed*(1 - energyLost);
+
+    public double determineLaunchSpeed(double distanceToTarget){
+        //declaration of % bounds
+        double range = initialUpperBound - initialLowerBound;
+        double launchPower = initialUpperBound;
+        double heightAtTargetDistance;
+
+        //Determine height at target distance
+     heightAtTargetDistance= getHeightAtTargetDistance(distanceToTarget,launchPower);
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:44]:
```suggestion
        heightAtTargetDistance= getHeightAtTargetDistance(distanceToTarget,launchPower);
```

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        else if(heightAtTargetDistance == targetHeight){
+            return maxLaunchSpeed*launchPower; // percent of max speed we should be going times the max speed 
+        }
+        //iteration
+        else{
+            for (int i = 0; i< 15; i++){//binary search...gets close enough to what we want
+                if(heightAtTargetDistance > targetHeight){//if our current height is greater than our target height raise lowerbound
+                    launchPower = launchPower - (range/(2*Math.pow(2,i))); //decreases launchPower
+                }
+                    else if(heightAtTargetDistance < targetHeight){
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:59]: needs some formatting love

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    }
+    private double getHeightAtTargetDistance(double distanceToTarget, double launchPower){
+        double horizontalSpeed;
+        double travelTime;
+        //if you have a triangle, with the hypnotonuse being the angled velocity vector, the adjacent being the horizontal velocity vector, and the opposite side being the height
+        //horizontal speed = adjacent side -> cos(angle) = a/h -> hcos(angle) = a -> (maxLaunchSpeed * launchPower)cos(angle) = a
+        horizontalSpeed = Math.cos(radians) * maxLaunchSpeed * launchPower;
+        //v = d/t -> vt=d -> t = d/v
+        travelTime = distanceToTarget/horizontalSpeed;;
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:80]:
```suggestion
        travelTime = distanceToTarget/horizontalSpeed;
```

Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:80]: Can you just declare the variables on line 74/75 down here?

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        //current distance = original distance +vt + .5at^2
+        double heightAtTargetDistance = launcherHeight + (Math.sin(radians)* maxLaunchSpeed * launchPower *travelTime) -(0.5 * Gravity *Math.pow(travelTime, 2 ));
+        return heightAtTargetDistance;
+    }
+
+
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:86]: get rid of these carriage returns please

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    private final double Gravity = 386.0886;
+    private final double maxRpm = 7640.0; 
+    private final double launcherHeight = 24.0; //inch
+    private final double energyLost = 0.66; 
+    private final double wheelCircumference = wheelDiameter*Math.PI;
+    private final double maxTangentialSpeed = ( wheelCircumference * (maxRpm /60));
+    private final double initialUpperBound = 1.0; //%100 speed
+    private final double initialLowerBound = 0.0;//%0 speed
+    private final double degrees = 58.0; 
+    private final double radians = Math.toRadians(degrees);
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:33]: nitpick: let's make final variables all caps

e.g. `private final double RADIANS = [....]`

Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:33]: Can we rename these variables to be more descriptive? This is the shooter angle right? So maybe something like `SHOOTER_ANGLE_DEGREES` and `SHOOTER_ANGLE_RADIANS`

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    private final double wheelDiameter = 4; //inch
+    private final double Gravity = 386.0886;
+    private final double maxRpm = 7640.0; 
+    private final double launcherHeight = 24.0; //inch
+    private final double energyLost = 0.66; 
+    private final double wheelCircumference = wheelDiameter*Math.PI;
+    private final double maxTangentialSpeed = ( wheelCircumference * (maxRpm /60));
+    private final double initialUpperBound = 1.0; //%100 speed
+    private final double initialLowerBound = 0.0;//%0 speed
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:31]: perhaps `INITIAL_UPPER_BOUND_POWER`? initialUpperBound and initialLowerBound aren't very descriptive

**Pull request merged**